invisible_cities/reco/psf_functions.py

```diff
@@ -0,0 +1,133 @@
+import numpy  as np
+import pandas as pd
+
+from typing import Tuple
+from typing import List
+
+from ..core.core_functions import shift_to_bin_centers
+
+def createPSF(pos    : Tuple[np.ndarray, ...],
+              charge : np.ndarray,
+              nbins  : int,
+              ranges : List[List[float]]
+              ) -> Tuple[np.ndarray, np.ndarray, List[np.ndarray]] :
+    """
+    Computes the point-spread (PSF) function of a given dataset.
+
+    Parameters
+    ----------
+    pos    : Hits relative position. Only tested for 2D.
+    charge : Hits SiPM charge normalized to the total peak charge.
+    nbins  : The number of bins in each dimension.
+    ranges : Range of the PSF in each dimension.
+
+    Returns
+    ----------
+    psf         : Point-spread function.
+    entries     : Number of entries per bin in the PSF.:
+    bin_centers : Bin centers of the PSF.
+    """
+
+    entries, edges = np.histogramdd(pos, nbins, range=ranges, normed=False)
+    sumC   , edges = np.histogramdd(pos, nbins, range=ranges, normed=False, weights=charge)
+    with np.errstate(divide='ignore', invalid='ignore'):
+        psf = np.nan_to_num(sumC/entries)
+
+    return psf, entries, [shift_to_bin_centers(edge) for edge in edges]
+
+
+def hdst_PSF_processing(dsts        : pd.DataFrame,
+                        ranges      : List[List[float]],
+                        sampleWidth : List[float]
+                        ) -> pd.DataFrame :
+    """
+    Adds the necessary info to a hits DST to create the PSF, namely the relative position and the normalized Q.
+
+    Parameters
+    ----------
+    dsts        : hits (1 SiPM per hit).
+    ranges      : range of the PSF in each dimension.
+    sampleWidth : Sampling distance of the hits.
+
+    Returns
+    ----------
+    hdst        : hits after processing to create PSF.
+    """
+    def AddVariableWeightedMean(df        : pd.DataFrame,
+                                varMean   : str,
+                                varWeight : str
+                                ) -> pd.DataFrame :
+        """
+        Adds the average of a variable weighted by another to a
+        grouped hits DST 'df' (grouped using groupby, by event id).
+
+        Parameters
+        ----------
+        df        : groupby by event and npeak.
+        varMean   : variable to be averaged.
+        varWeight : variable to be uses as the weight.
+
+        Returns
+        ----------
+        df        : dst with the weighted average.
+        """
+        df[varMean + 'peak'] =  (df[varMean]*df[varWeight]).sum()/df[varWeight].sum()
+        return(df)
```

Comment:
Chaining?

Reply:
You mean chaining the input df with the new column?

Reply:
I'm not saying anything about this function specifically, rather generally addressing your comment

> The function modifies the object, no need to return it

In general, even if a function mutates its input, you may want to return it anyway, in order to allow chaining operations on it. Probably most useful (well, prettiest) when applied to member functions rather than free functions. Compare:

````python
weapon = make_weapon()
weapon.safety_off()
weapon.fire()
smoking_gun = weapon.safety_on()
````

with 

````python
smoking_gun = make_weapon().safety_off().fire().safety_on()
````

With free functions it would be

````python
weapon = make_weapon()
safety_off(weapon)
fire(weapon)
smoking_gun = safety_on(weapon)
````
and

````python
smoking_gun = safety_on(fire(safety_off(make_weapon())
````

(Can't let that last one go by without a nod to the Haskell spellings of this:
````haskell
smoking_gun = safety_on (fire (safety_off make_weapon)) -- direct translation (spaces not necessary, but conventional)

smoking_gun = (safety_on . fire . safety_off) make_weapon -- function composition (spaces optional ... and in the following ones too)

smoking_gun = safety_on . fire . safety_off $ make_weapon -- function composition + application operator

smoking_gun = safety_on $ fire $ safety_off make_weapon -- application operator (essentially: avoid writing brackets)

smoking_gun = make_weapon & safety_off & fire & safety_on -- reverse application operator: piping
````
)

Reply:
I agree it can be useful, but my intuition when a function returns something tells me that it is a new object. Since we are already building functions with side effects (which in this case i think is safe because the scope of the object is rather well defined), I would like it at least to be obvious which is not from this `smoking_gun` example. That of course doesn't happen in Haskell.

Actually i just realized that in the review I used the word *need*. I actually meant that I would rather not do so for the aforementioned reason.

Reply:
Yup, it's a question of style and expectations. Your desire to stick to a certain style and associate certain expectations with it, is a perfectly valid choice with which I have a lot of sympathy. I'm merely pointing out that it's not the only way to do things, and that not everyone will share your expectations, at least not without documentation / education / agreement.

Continuing the Haskell asides: what is cool about Haskell is that it ensures that, in these sorts of cases, there can be no surprises; the type system tells you exactly what to expect, and ensures that the expectations will not by violated. In plain code it is guaranteed that nothing will be mutated, but by expressing your ideas in some monad of your choosing you switch to a different set of expectations and have them enforced by the language.

Reply:
>  a perfectly valid choice with which I have a lot of sympathy

Well, I hope, because it is your fault that I have that approach :laughing: 

> not everyone will share your expectations

Yes, that is a good point and I should defenitely keep it in mind :+1: 

> ensures that the expectations will not by violated.

Indeed. And I've found that more robust and easier to debug!

> but by expressing your ideas in some monad of your choosing you switch to a different set of expectations and have them enforced by the language.

I am still to understand monads... And I think I am still faraway from doing so.

Reply:
> a good point and I should defenitely keep it in mind 

Good, as it's my central point: we have built a certain culture in IC, which implicitly brings certain expectations with it. We need to put at least a little effort into maintaiting the culture and expectations, for our design choices to continue to make sense to collaborators as time passes. So your review comments like the one from which I started this digression are important to make, and I'm prodding it because I think that it is important that we remind ourselves why we use this style, and also to raise awareness that it is a choice we have made, rather than being god-given.

> And I've found that more robust and easier to debug!

Should I infer from this that you've been doing some Haskell on the side recently?

> I am still to understand monads... And I think I am still faraway from doing so.

In many ways they are less difficult than the fuss around them suggests. I'd be happy to help, if you want, but this is not the place for it.

Reply:
>  it is important that we remind ourselves why we use this style, and also to raise awareness that it is a choice we have made, rather than being god-given.

Noted! thanks!

> Should I infer from this that you've been doing some Haskell on the side recently?

Sporadically. I had two periods of time in which I dedicated a few hours a week to learn.  The progress in the second attempt was way more successful than in the first, so my hope is that when I try again it will be even better.

> I'd be happy to help, if you want

Thanks a lot. I *will* definitely take the offer. Maybe this will trigger the third Haskell campaign!